src/ipc/ipc4/helper.c

```diff
@@ -372,12 +372,16 @@ int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
 	irq_local_enable(flags);
 
 	ret = comp_bind(source, bu);
-	if (ret < 0)
-		return IPC4_INVALID_RESOURCE_ID;
+	if (ret < 0) {
+		ret = IPC4_INVALID_RESOURCE_ID;
+		goto free;
+	}
 
 	ret = comp_bind(sink, bu);
-	if (ret < 0)
-		return IPC4_INVALID_RESOURCE_ID;
+	if (ret < 0) {
```

Comment:
Still memory leak is possible - source has been bind, sink not, and procedure fail in the middle
not sure what happens (crash?) if buffer id freed when source bind was a success

Reply:
> FWIW I don't see a single `.bind()` implementation. Maybe time to remove it

@lyakh the mixinmixout uses the bind/unbind. I am about to remove it but not there yet

Reply:
@ranj063 you're right, thanks, I misinterpreted my grep results